File: lib/src/main/java/net/mm2d/preference/HeaderLoader.java
/*
 * Copyright (c) 2018 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.preference;

import android.content.Context;
import android.content.Intent;
import android.content.res.TypedArray;
import android.content.res.XmlResourceParser;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.XmlRes;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.util.Xml;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.util.List;

/**
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
class HeaderLoader {
    static void loadFromResource(
            @NonNull final Context context,
            @XmlRes final int resId,
            @NonNull final List<Header> target) {
        XmlResourceParser parser = null;
        try {
            parser = context.getResources().getXml(resId);
            loadFromResource(context, parser, target);
        } catch (final XmlPullParserException e) {
            throw new RuntimeException("Error parsing headers", e);
        } catch (final IOException e) {
            throw new RuntimeException("Error parsing headers", e);
        } finally {
            if (parser != null) {
                parser.close();
            }
        }
    }

    private static void loadFromResource(
            @NonNull final Context context,
            @NonNull final XmlResourceParser parser,
            @NonNull final List<Header> target)
            throws IOException, XmlPullParserException {
        final AttributeSet attrs = Xml.asAttributeSet(parser);
        while (true) {
            final int type = parser.next();
            if (type == XmlPullParser.END_DOCUMENT ||
                    type == XmlPullParser.START_TAG) {
                break;
            }
        }
        String nodeName = parser.getName();
        if (!"preference-headers".equals(nodeName)) {
            throw new RuntimeException(
                    "XML document must start with <preference-headers> tag; found"
                            + nodeName + " at " + parser.getPositionDescription());
        }
        final int outerDepth = parser.getDepth();
        while (true) {
            final int type = parser.next();
            if (type == XmlPullParser.END_DOCUMENT) {
                break;
            }
            if (type == XmlPullParser.END_TAG && parser.getDepth() <= outerDepth) {
                break;
            }
            if (type == XmlPullParser.END_TAG || type == XmlPullParser.TEXT) {
                continue;
            }

            nodeName = parser.getName();
            if ("header".equals(nodeName)) {
                target.add(parseHeaderSection(context, parser, attrs));
            } else {
                skipCurrentTag(parser);
            }
        }
    }

    private static Header parseHeaderSection(
            @NonNull final Context context,
            @NonNull final XmlResourceParser parser,
            @NonNull final AttributeSet attrs)
            throws IOException, XmlPullParserException {
        final Header header = new Header();
        final TypedArray sa = context.obtainStyledAttributes(attrs, R.styleable.PreferenceHeader);
        header.id = sa.getResourceId(R.styleable.PreferenceHeader_id, (int) PreferenceActivityCompatDelegate.HEADER_ID_UNDEFINED);
        TypedValue tv = sa.peekValue(R.styleable.PreferenceHeader_title);
        if (tv != null && tv.type == TypedValue.TYPE_STRING) {
            if (tv.resourceId != 0) {
                header.titleRes = tv.resourceId;
            } else {
                header.title = tv.string;
            }
        }
        tv = sa.peekValue(R.styleable.PreferenceHeader_summary);
        if (tv != null && tv.type == TypedValue.TYPE_STRING) {
            if (tv.resourceId != 0) {
                header.summaryRes = tv.resourceId;
            } else {
                header.summary = tv.string;
            }
        }
        tv = sa.peekValue(R.styleable.PreferenceHeader_breadCrumbTitle);
        if (tv != null && tv.type == TypedValue.TYPE_STRING) {
            if (tv.resourceId != 0) {
                header.breadCrumbTitleRes = tv.resourceId;
            } else {
                header.breadCrumbTitle = tv.string;
            }
        }
        header.iconRes = sa.getResourceId(R.styleable.PreferenceHeader_icon, 0);
        header.fragment = sa.getString(R.styleable.PreferenceHeader_fragment);
        sa.recycle();
        parseIntentSection(context, parser, attrs, header);
        return header;
    }

    private static void parseIntentSection(
            @NonNull final Context context,
            @NonNull final XmlResourceParser parser,
            @NonNull final AttributeSet attrs,
            @NonNull final Header header)
            throws IOException, XmlPullParserException {
        final Bundle curBundle = new Bundle();
        final int innerDepth = parser.getDepth();
        while (true) {
            final int type2 = parser.next();
            if (type2 == XmlPullParser.END_DOCUMENT) {
                break;
            }
            if (type2 == XmlPullParser.END_TAG && parser.getDepth() <= innerDepth) {
                break;
            }
            if (type2 == XmlPullParser.END_TAG || type2 == XmlPullParser.TEXT) {
                continue;
            }

            final String innerNodeName = parser.getName();
            switch (innerNodeName) {
                case "extra":
                    context.getResources().parseBundleExtra("extra", attrs, curBundle);
                    skipCurrentTag(parser);
                    break;
                case "intent":
                    header.intent = Intent.parseIntent(context.getResources(), parser, attrs);
                    break;
                default:
                    skipCurrentTag(parser);
                    break;
            }
        }
        if (curBundle.size() > 0) {
            header.fragmentArguments = curBundle;
        }
    }

    private static void skipCurrentTag(final XmlPullParser parser)
            throws IOException, XmlPullParserException {
        final int outerDepth = parser.getDepth();
        while (true) {
            final int type = parser.next();
            if (type == XmlPullParser.END_DOCUMENT) {
                break;
            }
            if (type == XmlPullParser.END_TAG && parser.getDepth() <= outerDepth) {
                break;
            }
        }
    }
}

File: lib/src/main/java/net/mm2d/preference/HeaderLoader.kt
/*
 * Copyright (c) 2018 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.preference

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.res.XmlResourceParser
import android.os.Bundle
import android.util.AttributeSet
import android.util.TypedValue
import android.util.Xml
import androidx.annotation.XmlRes
import androidx.core.content.res.use
import net.mm2d.preference.PreferenceActivityCompatDelegate.Companion.HEADER_ID_UNDEFINED
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.IOException

internal object HeaderLoader {
    fun loadFromResource(
        context: Context,
        @XmlRes resId: Int,
        target: MutableList<Header>
    ) {
        var parser: XmlResourceParser? = null
        try {
            parser = context.resources.getXml(resId)
            loadFromResource(context, parser, target)
        } finally {
            parser?.close()
        }
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun loadFromResource(
        context: Context,
        parser: XmlResourceParser,
        target: MutableList<Header>
    ) {
        val attrs = Xml.asAttributeSet(parser)
        @Suppress("ControlFlowWithEmptyBody")
        while (parser.next().let { it != XmlPullParser.END_DOCUMENT && it != XmlPullParser.START_TAG });
        if ("preference-headers" != parser.name) {
            throw RuntimeException(
                "XML document must start with <preference-headers> tag; found"
                    + parser.name + " at " + parser.positionDescription
            )
        }
        val startDepth = parser.depth
        while (true) {
            val type = parser.next()
            if (reachToEnd(type, parser.depth, startDepth)) {
                break
            }
            if (type == XmlPullParser.END_TAG || type == XmlPullParser.TEXT) {
                continue
            }
            if ("header" == parser.name) {
                target.add(parseHeaderSection(context, parser, attrs))
            } else {
                skipCurrentTag(parser)
            }
        }
    }

    @SuppressLint("Recycle")
    @Throws(IOException::class, XmlPullParserException::class)
    private fun parseHeaderSection(
        context: Context,
        parser: XmlResourceParser,
        attrs: AttributeSet
    ): Header {
        val header = Header()
        context.obtainStyledAttributes(attrs, R.styleable.PreferenceHeader).use {
            header.id = it.getResourceId(R.styleable.PreferenceHeader_id, HEADER_ID_UNDEFINED)
            header.setTitle(it.peekValue(R.styleable.PreferenceHeader_title))
            header.setSummary(it.peekValue(R.styleable.PreferenceHeader_summary))
            header.setBreadCrumbTitle(it.peekValue(R.styleable.PreferenceHeader_breadCrumbTitle))
            header.iconRes = it.getResourceId(R.styleable.PreferenceHeader_icon, 0)
            header.fragment = it.getString(R.styleable.PreferenceHeader_fragment)
        }
        parseIntentSection(context, parser, attrs, header)
        return header
    }

    private fun Header.setTitle(tv: TypedValue?) {
        if (tv?.type != TypedValue.TYPE_STRING) return
        if (tv.resourceId != 0) {
            titleRes = tv.resourceId
        } else {
            title = tv.string
        }
    }

    private fun Header.setSummary(tv: TypedValue?) {
        if (tv?.type != TypedValue.TYPE_STRING) return
        if (tv.resourceId != 0) {
            summaryRes = tv.resourceId
        } else {
            summary = tv.string
        }
    }

    private fun Header.setBreadCrumbTitle(tv: TypedValue?) {
        if (tv?.type != TypedValue.TYPE_STRING) return
        if (tv.resourceId != 0) {
            breadCrumbTitleRes = tv.resourceId
        } else {
            breadCrumbTitle = tv.string
        }
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun parseIntentSection(
        context: Context,
        parser: XmlResourceParser,
        attrs: AttributeSet,
        header: Header
    ) {
        val bundle = Bundle()
        val startDepth = parser.depth
        while (true) {
            val type = parser.next()
            if (reachToEnd(type, parser.depth, startDepth)) {
                break
            }
            if (type == XmlPullParser.END_TAG || type == XmlPullParser.TEXT) {
                continue
            }
            when (parser.name) {
                "extra" -> {
                    context.resources.parseBundleExtra("extra", attrs, bundle)
                    skipCurrentTag(parser)
                }
                "intent" -> header.intent = Intent.parseIntent(context.resources, parser, attrs)
                else -> skipCurrentTag(parser)
            }
        }
        if (bundle.size() > 0) {
            header.fragmentArguments = bundle
        }
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun skipCurrentTag(parser: XmlPullParser) {
        val startDepth = parser.depth
        @Suppress("ControlFlowWithEmptyBody")
        while (!reachToEnd(parser.next(), parser.depth, startDepth));
    }

    private fun reachToEnd(
        type: Int,
        currentDepth: Int,
        startDepth: Int
    ): Boolean = type == XmlPullParser.END_DOCUMENT ||
        type == XmlPullParser.END_TAG && currentDepth <= startDepth
}
